notification/notification-handler.es6

```diff
@@ -3,77 +3,134 @@
-		});
-	}, 100);
+		if (user.isGuest(userId)) {
+			let notes;
 
-	// Create notification events
-	core.on("text-dn", text => {
-		let userId = store.get("user");
+			try {
+				notes = JSON.parse(window.localStorage.getItem("notes"))[userId];
+			} catch (e) {
+				console.log("Failed to load cached notes", e);
+			}
 
-		if (text.from === userId) {
-			return;
+			if (Array.isArray(notes)) {
+				core.emit("setstate", { notes });
+			}
 		}
+	}
+
+	core.on("init-dn", () => {
 
-		if (text.tags && text.tags.indexOf("hidden") > -1) {
-			return;
+		loadCache();
+
+		if (!user.isGuest(store.get("user"))) {
+			// Delete cached notes if not guest
+			try {
+				window.localStorage.removeItem("notes");
+			} catch (e) {
+				console.log("Failed to clear cached notes", e);
+			}
+
+			core.emit("getNotes", {}, (err, res) => {
+				let notes;
+
+				if (err) {
+					notes = [];
+				} else {
+					notes = res.results;
+				}
+
+				if (Array.isArray(notes)) {
+					core.emit("setstate", { notes });
+				}
+			});
 		}
+	}, 100);
+
+	window.addEventListener("storage", loadCache, false);
+
+	// Update notifications cache if guest
+	core.on("statechange", changes => {
+		let userId = store.get("user");
 
-		// FIXME: ugly hack until messages in hidden threads are no longer received
-		if (text.thread) {
-			let thread = store.get("indexes", "threadsById", text.thread);
+		if (user.isGuest(userId) && changes.notes) {
+			let notes = { [userId]: store.get("notes") };
 
-			if (thread && thread.tags.indexOf("thread-hidden") > -1) {
-				return;
+			try {
+				localStorage.setItem("notes", JSON.stringify(notes));
+			} catch (e) {
+				console.log("Failed to cache notes", e);
 			}
 		}
+	}, 100);
 
-		let subtype;
+	let actions = [ "note-dn", "note-up" ];
 
-		if (text.mentions && text.mentions.indexOf(userId) > -1) {
-			subtype = "mention";
-		} else {
+	actions.forEach(action => {
+		core.on(action, note => {
 			let roomId = store.get("nav", "room"),
 				mode = store.get("nav", "mode");
 
-			if (text.id === text.thread) {
-				let roomId = store.get("nav", "room"),
-					mode = store.get("nav", "mode");
-
-				if (roomId === text.to && mode === "room" && !user.isAdmin(userId, roomId)) {
-					return;
-				}
-
-				subtype = "thread";
-			} else {
+			// TODO: figure out a better way
+			if (mode === "chat") {
 				let threadId = store.get("nav", "thread");
 
-				if (roomId === text.to && mode === "chat" && ((threadId && threadId === text.thread) || (!threadId && !text.thread))) {
+				if (threadId) {
+					if (note.group === roomId + "/" + threadId) {
+						return;
+					}
+				} else {
+					if (note.group === roomId) {
+						return;
+					}
+				}
+			} else if (mode === "room") {
+				if (note.group === roomId) {
 					return;
 				}
+			}
 
-				subtype = "text";
+			let notes = store.get("notes").slice(0);
+
+			notes.push(note);
+
+			core.emit("setstate", { notes });
+		}, 10);
+	});
+
+	core.on("setstate", changes => {
```

Comment:
Please move this to a rule `dismissNotificationsOnNav` to stay consistent and remove business logic from components.